tstest/natlab/firewall.go

```diff
@@ -5,26 +5,67 @@
 package natlab
 
 import (
+	"fmt"
 	"sync"
 	"time"
 
 	"inet.af/netaddr"
 )
 
-type session struct {
+// EndpointDependence expresses
+type FirewallSelectivity int
+
+const (
+	// APDF specifies a destination address-and-port dependent
+	// firewall. Outbound traffic to an ip:port authorizes traffic
+	// from that ip:port exactly, and nothing else.
+	APDF FirewallSelectivity = iota
+	// ADF specifies a destination address dependent firewall. Once
+	// outbound traffic has been seen to an IP address, that IP
+	// address can talk back from any port.
+	ADF
+	// EIF specifies a destination endpoint independent firewall. Once
+	// outbound traffic has been seen from a source, anyone can talk
+	// back to that source.
+	EIF
+)
+
+type fwKey struct {
 	src netaddr.IPPort
 	dst netaddr.IPPort
 }
 
+func (s FirewallSelectivity) key(src, dst netaddr.IPPort) fwKey {
+	k := fwKey{src: src}
+	switch s {
+	case EIF:
+	case ADF:
+		k.dst.IP = dst.IP
+	case APDF:
+		k.dst = dst
+	default:
+		panic(fmt.Sprintf("unknown firewall selectivity %v", s))
+	}
+	return k
+}
+
+// Firewall is a simple stateful firewall that allows all outbound
+// traffic and filters inbound traffic based on recently seen outbound
+// traffic. Its HandlePacket method should be attached to a Machine to
+// give it a stateful firewall.
 type Firewall struct {
-	// TrustedInterface is the interface that's allowed to send
-	// anywhere. All other interfaces can only respond to traffic from
-	// TrustedInterface.
-	TrustedInterface *Interface
 	// SessionTimeout is the lifetime of idle sessions in the firewall
 	// state. Packets transiting from the TrustedInterface reset the
 	// session lifetime to SessionTimeout.
 	SessionTimeout time.Duration
+	// Selectivity specifies how precisely return traffic must match
```

Comment:
what does the zero value mean? AddressPort? Perhaps those consts should start at 1, not 0, so we can tell when the zero value is in use if you want different behavior? Or just document what you mean by zero here.

Reply:
Documented the default. I'm keeping the zero value as AddressAndPortDependentFirewall, because in practice that's the config we'll mostly want.